app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/vpn/service/TrackerBlockingVpnService.kt

```diff
@@ -167,121 +167,141 @@ class TrackerBlockingVpnService : VpnService(), CoroutineScope by MainScope(), V
             }
             ACTION_RESTART_VPN -> {
-                restartRequested = true
-                launch { stopVpn(VpnStopReason.RESTART) }
+                launch(serviceDispatcher) { startVpn(tunInterface) }
             }
-            else -> logcat(LogPriority.ERROR) { "Unknown intent action: $action" }
+            else -> logcat(ERROR) { "Unknown intent action: $action" }
         }
 
         return returnCode
     }
 
-    private suspend fun startVpn() = withContext(Dispatchers.IO) {
+    private suspend fun startVpn(currentTunFd: ParcelFileDescriptor?) = withContext(serviceDispatcher) {
+        fun updateNetworkStackUponRestart() {
+            logcat { "VPN log: updating the networking stack" }
+            logcat { "VPN log: CURRENT network ${vpnNetworkStack.name}" }
+            // stop the current networking stack
+            vpnNetworkStack.onStopVpn(VpnStopReason.RESTART)
+            vpnNetworkStack.onDestroyVpn()
+            // maybe we have changed the networking stack
+            vpnNetworkStack = vpnNetworkStackProvider.provideNetworkStack()
+            vpnNetworkStack.onCreateVpnWithErrorReporting()
+            logcat { "VPN log: NEW network ${vpnNetworkStack.name}" }
+        }
+
         logcat { "VPN log: Starting VPN" }
+        val restarting = currentTunFd != null
 
         synchronized(startVpnLock) {
             val currStateStats = vpnServiceStateStatsDao.getLastStateStats()
             // We don't check for ENABLED state here because it break when the app is updated
             // (See: https://app.asana.com/0/488551667048375/1203621692416589/f)
-            if (currStateStats?.state == ENABLING) {
+            if (!restarting && currStateStats?.state == ENABLING) {
                 // Sometimes onStartCommand gets called twice - this is a safety rail against that
-                logcat(LogPriority.WARN) { "VPN is already being started, abort" }
+                logcat(LogPriority.WARN) { "VPN log: is already being started, abort" }
                 return@withContext
             }
 
             vpnServiceStateStatsDao.insert(createVpnState(state = ENABLING))
 
-            vpnServiceCallbacksPluginPoint.getPlugins().forEach {
-                logcat { "VPN log: onVpnStarting ${it.javaClass} callback" }
-                it.onVpnStarting(this)
+            if (!restarting) {
+                vpnServiceCallbacksPluginPoint.getPlugins().forEach {
+                    logcat { "VPN log: onVpnStarting ${it.javaClass} callback" }
+                    it.onVpnStarting(this)
+                }
+            } else {
+                logcat { "VPN log: skipping service callbacks while restarting" }
             }
         }
 
+        // Create a null route tunnel so that leaks can't scape
+        createNullRouteTempTunnel()
+        // let Android time to create the tunnel before we possibly update the networks that will close the old tunnel
+        delay(100)
+        currentTunFd?.let {
```

Comment:
We are not closing this tun anymore. Do we just rely that it will be killed at some point?

Reply:
Yeah, there's no need to close it because the system re-routes the traffic as soon as new one is created. Plus then we don't need to worry about closing it too soon such that it creates leakages.